core/src/main/scala/org/apache/spark/sql/sources/SparkOptimizations.scala

```diff
@@ -0,0 +1,410 @@
+        newPlan transformUp {
+          case q: LogicalPlan =>
+            q transformExpressionsUp {
+              case a: AttributeReference => newAttributesMap.getOrElse(a, a)
+            }
+        }
+      } else {
+        newPlan
+      }
+    }
+  }
+
+  private def fetchIndexes(table: LogicalPlan) = table.collect {
+    case l@LogicalRelation(p: ParentRelation, _, _) =>
+      val catalog = snappySession.sessionCatalog
+      (l.asInstanceOf[LogicalPlan], p.getDependents(catalog).map(idx =>
+        catalog.lookupRelation(catalog.newQualifiedTableName(idx))))
+    case s@SubqueryAlias(alias, child@LogicalRelation(p: ParentRelation, _, _)) =>
+      val catalog = snappySession.sessionCatalog
+      (s.asInstanceOf[LogicalPlan], p.getDependents(catalog).map(idx =>
+        catalog.lookupRelation(catalog.newQualifiedTableName(idx))))
+  }
+
+  object HasColocatedEntities {
+
+    type ReturnType = (
+        Seq[(BaseColumnFormatRelation, BaseColumnFormatRelation)], // left & right relation
+            Seq[(LogicalPlan, LogicalPlan)], // Seq[left (tablePlan, indexPlan)]
+            Seq[(LogicalPlan, LogicalPlan)] // Seq[right (tablePlan, indexPlan)]
+        )
+
+    def unapply(tables: (LogicalPlan, LogicalPlan)):
+    Option[ReturnType] = {
+      val (left, right) = tables
+
+      /** now doing a one-to-one mapping of lefttable and its indexes with
+        * right table and its indexes. Following example explains the combination
+        * generator.
+        *
+        * val l = Seq(1, 2, 3)
+        * val r = Seq(4, 5)
+        * l.zip(Seq.fill(l.length)(r)).flatMap {
+        * case (leftElement, rightList) => rightList.flatMap { e =>
+        * Seq((leftElement, e))
+        * }
+        * }.foreach(println)
+        * will output :
+        * (1, 4)
+        * (1, 5)
+        * (2, 4)
+        * (2, 5)
+        * (3, 4)
+        * (3, 5)
+        *
+        */
+      val leftRightEntityMapping = for {
+        (leftTable, leftIndexes) <- fetchIndexes(left)
+        (rightTable, rightIndexes) <- fetchIndexes(right)
+        leftSeq = Seq(leftTable) ++ leftIndexes
+        rightSeq = Seq(rightTable) ++ rightIndexes
+      } yield {
+        leftSeq.zip(Seq.fill(leftSeq.length)(rightSeq)) flatMap {
+          case (leftElement, rightList) => rightList.flatMap(e => Seq((leftElement, e)))
+        }
+      }
+
+      // right now not expecting multiple tables in left & right hand side.
+//      assert(leftRightEntityMapping.size <= 1)
+
+      val mappings = leftRightEntityMapping.flatMap { mappedElements =>
+        val (leftTable, rightTable) = mappedElements(0) // first pairing is always (table, table)
+        for {
+
+          (leftPlan, rightPlan) <- mappedElements
+
+          leftRelation = unwrapBaseColumnRelation(leftPlan) if leftRelation.nonEmpty
+          rightRelation = unwrapBaseColumnRelation(rightPlan) if rightRelation.nonEmpty
+
+          if isColocated(leftRelation.get, rightRelation.get)
+        } yield {
```

Comment:
Since we are expecting only one left and right replacement, can we use a construct other than for-yield?